lib/ex_rlp/decode.ex

```diff
@@ -3,83 +3,117 @@ defmodule ExRLP.Decode do
+
+    decode_item(new_tail, new_result)
   end
 
-  # Decodes long length-binary?
   defp decode_item(<<(<<be_size_prefix>>), tail::binary>>, result) when be_size_prefix < 192 do
     {new_item, new_tail} = decode_long_binary(be_size_prefix, tail, 183)
 
-    decode_item(new_tail, [new_item | result])
+    new_result = add_new_item(result, new_item)
+
+    decode_item(new_tail, new_result)
+  end
+
+  defp decode_item(<<(<<be_size_prefix>>), tail::binary>>, result) when be_size_prefix == 192 do
+    new_item = []
+
+    new_result = add_new_item(result, new_item)
+
+    decode_item(tail, new_result)
+  end
+
+  defp decode_item(<<(<<prefix>>), tail::binary>>, result) when prefix <= 247 do
+    {list, new_tail} = decode_medium_binary(prefix, tail, 192)
+
+    new_result = add_decoded_list(result, list)
+
+    decode_item(new_tail, new_result)
+  end
+
+  defp decode_item(<<(<<be_size_prefix>>), tail::binary>>, result) do
+    {list, new_tail} = decode_long_binary(be_size_prefix, tail, 247)
+
+    new_result = result |> add_decoded_list(list)
+
+    decode_item(new_tail, new_result)
   end
 
-  # Decodes an empty list
-  defp decode_item(<<(<<be_size_prefix>>), tail::binary>>, result)
-       when be_size_prefix == 192 do
-    decode_item(tail, [[] | result])
+  @spec add_new_item(ExRLP.t(), ExRLP.t()) :: ExRLP.t()
+  def add_new_item(nil, new_item) do
+    new_item
   end
 
-  defp decode_item(<<(<<prefix>>), tail::binary>>, result) do
-    {list_bin, new_tail} =
-      if prefix <= 247 do
-        decode_medium_binary(prefix, tail, 192)
-      else
-        decode_long_binary(prefix, tail, 247)
-      end
+  def add_new_item(result, new_item) do
+    [new_item | result]
+  end
+
+  @spec add_decoded_list(ExRLP.t(), binary()) :: ExRLP.t()
+  defp add_decoded_list(nil, rlp_list_binary) do
+    rlp_list_binary
+    |> decode_item([])
+    |> Enum.reverse()
+  end
 
-    next_list = decode_item(list_bin, [])
+  defp add_decoded_list(result, rlp_list_binary) do
+    list_items = decode_item(rlp_list_binary, [])
 
-    decode_item(new_tail, [next_list | result])
+    [list_items | result]
   end
 
   @spec decode_medium_binary(integer(), binary(), integer()) :: {binary(), binary()}
   defp decode_medium_binary(length_prefix, tail, prefix) do
     item_length = length_prefix - prefix
-    <<item::binary-size(item_length), new_tail::binary()>> = tail
+    <<item::binary-size(item_length), new_tail::binary>> = tail
 
     {item, new_tail}
   end
 
   @spec decode_long_binary(integer(), binary(), integer()) :: {binary(), binary()}
   defp decode_long_binary(be_size_prefix, tail, prefix) do
     be_size = be_size_prefix - prefix
-    <<be::binary-size(be_size), data::binary()>> = tail
+    <<be::binary-size(be_size), data::binary>> = tail
 
     item_length = :binary.decode_unsigned(be)
-
-    <<item::binary-size(item_length), new_tail::binary()>> = data
+    <<item::binary-size(item_length), new_tail::binary>> = data
```

Comment:
would there be an option to extract this into a function head?
if you compile this with _bin_opt_info_ flags you get informed that 
NOT OPTIMIZED: sub binary is used or returned 
you can read more about this here http://erlang.org/documentation/doc-6.0/doc/efficiency_guide/binaryhandling.html

Reply:
I don't think so because length is not constant